index.js

```diff
@@ -70,20 +71,111 @@ function format(text, opts, addAlignmentSize) {
   return str;
 }
 
-function formatRange(text, opts) {
+function findNodeByOffset(node, offset) {
+  const start = util.locStart(node);
+  const end = util.locEnd(node);
+  if (start <= offset && offset <= end) {
+    for (const childNode of comments.getSortedChildNodes(node)) {
+      const childResult = findNodeByOffset(childNode, offset);
+      if (childResult) {
+        return childResult;
+      }
+    }
+
+    if (isSourceElement(node)) {
+      return node;
+    }
+  }
+}
+
+// See https://www.ecma-international.org/ecma-262/5.1/#sec-A.5
+function isSourceElement(node) {
+  if (node == null) {
+    return false;
+  }
+  switch (node.type) {
+    case "FunctionDeclaration":
+      return true;
+    // Statements
+    case "BlockStatement":
+    case "BreakStatement":
+    case "ContinueStatement":
+    case "DebuggerStatement":
+    case "DoWhileStatement":
+    case "EmptyStatement":
+    case "ExpressionStatement":
+    case "ForInStatement":
+    case "ForStatement":
+    case "IfStatement":
+    case "LabeledStatement":
+    case "ReturnStatement":
+    case "SwitchStatement":
+    case "ThrowStatement":
+    case "TryStatement":
+    case "VariableDeclaration":
+    case "WhileStatement":
+    case "WithStatement":
+      return true;
+  }
+  return false;
+}
+
+function calculateRange(text, opts, ast) {
+  // Contract the range so that it has non-whitespace characters at its endpoints.
+  // This ensures we can format a range that doesn't end on a node.
+  const rangeStringOrig = text.slice(opts.rangeStart, opts.rangeEnd);
+  const rangeStringOrigReversed = rangeStringOrig.split("").reverse().join("");
```

Comment:
Nooooo :( You don't want to allocate an array with every element, reverse it, and recreate a string to know the offset of the first and last non whitespace. Especially when the string here can be a 10k lines file.

```js
for (var lastIndex = opts.rangeEnd; lastIndex >= opts.rangeStart; --lastIndex) {
  if (text[lastIndex].match(/\S/)) {
    break;
  }
}
```

You can use `.search` for the first one.